micrometer-core/src/main/java/io/micronaut/configuration/metrics/micrometer/intercept/TimedInterceptor.java

```diff
@@ -75,86 +79,123 @@ protected TimedInterceptor(MeterRegistry meterRegistry) {
-                    exceptionClass = e.getClass().getSimpleName();
-                    throw e;
-                } finally {
-                    if (stop) {
-
-                        stopTimed(metricName, sample, exceptionClass, metadata);
-                    }
-                }
+
+                return Publishers.convertPublisher(single, javaReturnType);
+            } else {
+                AtomicReference<String> exceptionClass = new AtomicReference<>("none");
+                Flowable<?> flowable = Publishers.convertPublisher(result, Flowable.class);
+                flowable = flowable.doOnRequest(n -> sample.set(Timer.start(meterRegistry)))
+                        .doOnError(throwable -> exceptionClass.set(throwable.getClass().getSimpleName()))
+                        .doOnComplete((Action) () -> {
+                            final Timer.Sample s = sample.get();
+                            if (s != null) {
+                                stopTimed(
+                                        metricName,
+                                        s,
+                                        exceptionClass.get(),
+                                        metadata);
+                            }
+                        });
+                return Publishers.convertPublisher(flowable, javaReturnType);
             }
         }
-        return context.proceed();
+    }
+
+    private Object handleKotlinCoroutine(MethodInvocationContext<Object, Object> context, final AnnotationMetadata metadata, final String metricName) {
+        final Timer.Sample sample = Timer.start(meterRegistry);
+        final Object result = context.proceed();
+        // is it suspended?
+        if (result == null || !result.getClass().getName().equals("kotlin.coroutines.intrinsics.CoroutineSingletons")) {
+            // not suspended
+            stopTimed(
+                    metricName,
+                    sample,
+                    "none",
+                    metadata);
+        } else {
+            // last argument
+            final String[] argumentNames = context.getArgumentNames();
+            final MutableArgumentValue<?> argument = context.getParameters().get(argumentNames[argumentNames.length - 1]);
+            if (argument.getValue().getClass().getName().equals("io.micronaut.http.bind.binders.CustomContinuation")) {
+                final CompletableFuture<?> future = ((Supplier<CompletableFuture<?>>) argument.getValue()).get();
+                future.whenComplete((lastResult, exception) ->
+                    stopTimed(
+                            metricName,
+                            sample,
+                            exception == null ? "none" : exception.getClass().getSimpleName(),
+                            metadata)
+                );
+            }
+        }
+        return result;
+    }
+
+    private static boolean isKotlinContinuation(MethodInvocationContext<Object, Object> context) {
```

Comment:
I believe for this you can just call `context.getExecutableMethod().isSuspend()`

Reply:
Done